### PR TITLE
8336289: Obliterate most references to _snprintf in the Windows JDK

### PR DESCRIPTION
**Pull request by TheShermanTanker**
snprintf has been available for all officially and unofficially supported compilers for Windows, Visual Studio since version 2015 and gcc since, well, forever. snprintf is conforming to C99 since the start when compiling using gcc, and 2015 when using Visual Studio. Since it conforms to C99 and provides better semantics than _snprintf, and is not compiler specific, we should use it in most places where we currently use _snprintf. This makes JDK code where this is used more robust due to the null terminating guarantee of snprintf. Since most of the changes are extremely simple, I've included all of them hoping to get this all done in one shot. The only places _snprintf is not replaced is places where I'm not sure whether the code is ours or not (As in, imported source code for external libraries). Note that the existing checks in place for the size of the characters written still work, as the return value of snprintf works mostly the same as _snprintf. The only difference is the null terminating character at the end and the returning of the number of written characters if the buffer was terminated early, as seen here: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170

Reviews Required: 2 for HotSpot, jdk.hotspot.agent, and jdk.jdwp.agent, 1 for awt and jdk.accessibility, 1 for java.base, 1 for security, 1 for jdk.management(?)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336289](https://bugs.openjdk.org/browse/JDK-8336289): Obliterate most references to _snprintf in the Windows JDK (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**) Review applies to [a0477b8b](https://git.openjdk.org/jdk/pull/20153/files/a0477b8b83f908c3e5c5df730cafa4d99c550406)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) Review applies to [a0477b8b](https://git.openjdk.org/jdk/pull/20153/files/a0477b8b83f908c3e5c5df730cafa4d99c550406)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20153/head:pull/20153` \
`$ git checkout pull/20153`

Update a local copy of the PR: \
`$ git checkout pull/20153` \
`$ git pull https://git.openjdk.org/jdk.git pull/20153/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20153`

View PR using the GUI difftool: \
`$ git pr show -t 20153`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20153.diff">https://git.openjdk.org/jdk/pull/20153.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20153#issuecomment-2224910546)

### CODE REVIEW

**Comment by kimbarrett** [comment]
This should be using `os::snprintf` rather than `snprintf`.  Rationale is in the comment here:
https://github.com/openjdk/jdk/blob/1f6e106b45e5109224e13d70f1a40c9e666ec2ab/src/hotspot/share/runtime/os.cpp#L118-L126

And yes, I know there are lots of bare uses of snprintf (about 125?), including in shared code.  That's why it isn't
currently in the forbidden list.  There's some cleanup to do there.

**Comment by TheShermanTanker** [comment]
> This should be using `os::snprintf` rather than `snprintf`. Rationale is in the comment here:
> 
> https://github.com/openjdk/jdk/blob/1f6e106b45e5109224e13d70f1a40c9e666ec2ab/src/hotspot/share/runtime/os.cpp#L118-L126
> 
> And yes, I know there are lots of bare uses of snprintf (about 125?), including in shared code. That's why it isn't currently in the forbidden list. There's some cleanup to do there.

Ah, I'd assumed that the places where bare \_snprintf was used had a reason for using it directly. I'll change all the usages in this Pull Request to use os::snprintf

**Comment by TheShermanTanker** [comment]
I would add a FORBID\_C\_FUNCTION for \_snprintf for Windows, but unfortunately that would be completely useless since there's no way to restrict methods on Visual Studio

File: src/jdk.jdwp.agent/windows/native/libjdwp/util_md.h
Comment (kimbarrett) [inline src/jdk.jdwp.agent/windows/native/libjdwp/util_md.h:32]: This change has nothing to do with _sprintf.  Not sure why it's being made here.

File: src/jdk.management/windows/native/libmanagement_ext/OperatingSystemImpl.c
Comment (kimbarrett) [inline src/jdk.management/windows/native/libmanagement_ext/OperatingSystemImpl.c:54]: Similarly, his change has nothing to do with _sprintf.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/jdk.jdwp.agent/windows/native/libjdwp/util_md.h
Comment (TheShermanTanker) [reply src/jdk.jdwp.agent/windows/native/libjdwp/util_md.h:32]: It was a small change, so I thought I could make it out of convenience. I'll switch it out to a separate changeset

Comment (kimbarrett) [reply src/jdk.jdwp.agent/windows/native/libjdwp/util_md.h]: There are lots of uses of those type names.  If they are going to be cleaned up, I'd prefer that get done as
as separate task, so I don't need to think about whether it's okay to do so while in the context of this change.
I don't know if __int64 == long long (probably is), but I'm pretty sure I remember seeing a comment on some
use of __int32 suggesting it was not the same as what someone expected.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
Okay for HotSpot, jdk.hotspot.agent, jdk.jdwp.agent and jdk.management.

Thanks

**Comment by TheShermanTanker** [comment]
Yeah, I do unfortunately need reviewers for the sections that haven't been approved yet. Taking this reminder as an opportunity to bump this Pull Request :)

**Review by prrace — APPROVED** [review]
AWT + A11Y

**Comment by TheShermanTanker** [comment]
Alright! Just need approval for java.base and security, then we should be good to go

**Review by jaikiran — APPROVED** [review]
Hello Julian,
> Just need approval for java.base and security

Daniel has approved this PR and I think that should take care of java.base review. In any case, I had a look at the java.base change and it looks OK to me.

A several of these files will need a copyright year change. Could you update those files to use 2024?

**Comment by TheShermanTanker** [comment]
I was going to ask for a security review, but I decided to only do that after I change the copyright years in the files that need it. Well, when I have the time that is...

**Comment by TheShermanTanker** [comment]
ALright, never mind, I've gone ahead with the copyright anyway, and in the process also fought my OCD and given the snprintf calls proper formatting too. Now awaiting re\-review from all, as well as the missing review from security

**Review by jaikiran — APPROVED** [review]

**Review by dholmes-ora — APPROVED** [review]
Still good for me. Thanks

**Review by kimbarrett — APPROVED** [review]
Still good.

**Review by seanjmullan — APPROVED** [review]
The change to j2secmod_md.c looks fine to me.

**Comment by TheShermanTanker** [comment]
Awesome! Will wait a while before integrating so objections, if any, can be raised. If no objections are raised by tomorrow morning I'll proceed with integration

**Comment by jaikiran** [comment]
> If no objections are raised by tomorrow morning I'll proceed with integration

I've run this PR against latest mainline in our CI for tier1, tier2 and tier3 and there were no failures. So yes, I think you can go ahead with the integration.

**Comment by TheShermanTanker** [comment]
Thanks everyone for the reviews!

**Comment by kevinjwalls** [comment]
This change is implicated in the regression JDK-8350820 in the Windows management/cpu usage functions.

**Comment by TheShermanTanker** [comment]
That's not good. I cannot view JDK\-8350820 however

**Comment by TheShermanTanker** [comment]
Is the issue in HotSpot, or another component?

**Comment by kevinjwalls** [comment]
Issue recently logged, was marked confidential, will update...

**Comment by kevinjwalls** [comment]
Quick fix is PR https://github.com/openjdk/jdk/pull/23832 taking care of it in JDK 25, and there should be a backport to JDK 24 first update.  I will come back with a further update in JDK-8350939 to reinstate snprintf.

**Pull request closed without merging**
